### PR TITLE
SCHED-511: Event Logic on GraphQL middleware 

### PR DESCRIPTION
**Pull request by stroncod**

### CODE REVIEW

**Review by sraaphorst — COMMENTED** [review]
Good progress here.

One thing I noticed that I didn't see before is the `NewWeatherChange` class and this line in particular:

```python
        c = Conditions(cc=CloudCover[self.new_CC.name] if self.new_CC else None,
                       sb=SkyBackground[self.new_SB.name] if self.new_SB else None,
                       wv=WaterVapor[self.new_WV.name] if self.new_WV else None,
                       iq=ImageQuality[self.new_IQ.name] if self.new_IQ else None)
```
I'm surprised that this isn't causing warnings, since the members of `Conditions` aren't `Optional`, so we shouldn't be able to set them to `None` without causing issues. I see what you're trying to do here, but I think a weather record with full info will come in rather than just waht has changed? In any case, we don't have to worry about this now, but I just wanted to bring it up to keep it in our minds.

File: scheduler/core/statscalculator/__init__.py [hunk 1]
```diff
@@ -18,47 +20,51 @@ class StatCalculator:
...
+
+                        all_programs_scores.setdefault(program.id, 0)
+                        all_programs_scores[program.id] += visit.score
+                        completion_fraction[program.band] += 1
+
+                        # Calculate altitude data
+                        ti = collector.get_target_info(visit.obs_id)
+                        end_time_slot = visit.start_time_slot + visit.time_slots
+                        values = ti[night_idx].alt[visit.start_time_slot: end_time_slot]
+                        alt_degs = [val.dms[0] + (val.dms[1] / 60) + (val.dms[2] / 3600) for val in values]
```
[the remark this answers is not in the document]
Comment (sraaphorst) [reply scheduler/core/statscalculator/__init__.py:61]: I think maybe `dtsex2dec` in `lucupy.timeutils` might do what you want?

File: scheduler/core/builder/builder.py [hunk 2]
```diff
@@ -32,6 +32,12 @@ def build_collector(self,
                         blueprint: CollectorBlueprint) -> Collector:
         # TODO: Removing sources from Collector I think it was an idea
         # TODO: we might want to implement so all these are static methods.
+
+        if isinstance(self.__class__, ValidationBuilder):
+            collector = Collector(start, end, sites, semesters, self.sources, *blueprint)
```
[the remark this answers is not in the document]
Comment (stroncod) [reply scheduler/core/builder/builder.py:37]: ValidationBuilder is used in the `run_` scripts so I didn't want it to interfere with those as the method is called outside but I need to call it inside the builder.

File: scheduler/core/statscalculator/__init__.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (stroncod) [reply scheduler/core/statscalculator/__init__.py:61]: I'll take a look

[the remark this answers is not in the document]
Comment (stroncod) [reply scheduler/core/statscalculator/__init__.py:61]: Those methods are not Astropy friendly but there are similar. I guess we can either merged the two or keep the logics separate.

[the remark this answers is not in the document]
Comment (sraaphorst) [reply scheduler/core/statscalculator/__init__.py:61]: Yep... let's not change it, then. It's not a big deal.

File: scheduler/core/eventsqueue/nightchanges.py [hunk 3]
```diff
@@ -2,7 +2,7 @@
 # For license information see LICENSE or https://opensource.org/licenses/BSD-3-Clause
 
 from dataclasses import dataclass, field
-from typing import Dict, List
+from typing import Dict, List, Optional
```
Comment (sraaphorst) [inline scheduler/core/eventsqueue/nightchanges.py:5]: I don't think we need `Optional` here... doesn't look like it's being used?

**Pull request merged**